File: iOS/Crisis Connect/Features/Profile/MedicalInfoStore.swift
//
//  MedicalInfoStore.swift
//  Crisis Connect
//
//  The user's optional emergency medical details (blood type, allergies, medication, note).
//  Deliberately LOCAL + BLE-ONLY: shared over the encrypted SOS peer_info with a verified
//  rescuer during an emergency, never uploaded to the cloud. Wire format matches Android's
//  MedicalInfoStore / BlePeerIdentityUtils ("medical": {blood, allergies, meds, notes}).
//

import Foundation

struct MedicalInfo: Equatable {
    var bloodType: String = ""
    var allergies: String = ""
    var medication: String = ""
    var notes: String = ""

    var isEmpty: Bool {
        bloodType.isBlank && allergies.isBlank && medication.isBlank && notes.isBlank
    }

    func sanitized() -> MedicalInfo {
        MedicalInfo(
            bloodType: String(bloodType.trimmingCharacters(in: .whitespacesAndNewlines)
                .prefix(MedicalInfoStore.maxBloodLength)),
            allergies: String(allergies.trimmingCharacters(in: .whitespacesAndNewlines)
                .prefix(MedicalInfoStore.maxFieldLength)),
            medication: String(medication.trimmingCharacters(in: .whitespacesAndNewlines)
                .prefix(MedicalInfoStore.maxFieldLength)),
            notes: String(notes.trimmingCharacters(in: .whitespacesAndNewlines)
                .prefix(MedicalInfoStore.maxFieldLength))
        )
    }

    /// The `"medical"` object for the peer_info payload; nil when everything is blank.
    func toPeerInfoJSON() -> [String: Any]? {
        let sanitized = sanitized()
        guard !sanitized.isEmpty else { return nil }
        var json: [String: Any] = [:]
        if !sanitized.bloodType.isBlank { json["blood"] = sanitized.bloodType }
        if !sanitized.allergies.isBlank { json["allergies"] = sanitized.allergies }
        if !sanitized.medication.isBlank { json["meds"] = sanitized.medication }
        if !sanitized.notes.isBlank { json["notes"] = sanitized.notes }
        return json.isEmpty ? nil : json
    }
}

enum MedicalInfoStore {
    static let maxBloodLength = 8
    static let maxFieldLength = 200

    private static let bloodKey = "medical.bloodType"
    private static let allergiesKey = "medical.allergies"
    private static let medicationKey = "medical.medication"
    private static let notesKey = "medical.notes"

    static func load(userDefaults: UserDefaults = .standard) -> MedicalInfo {
        MedicalInfo(
            bloodType: userDefaults.string(forKey: bloodKey) ?? "",
            allergies: userDefaults.string(forKey: allergiesKey) ?? "",
            medication: userDefaults.string(forKey: medicationKey) ?? "",
            notes: userDefaults.string(forKey: notesKey) ?? ""
        )
    }

    static func save(_ info: MedicalInfo, userDefaults: UserDefaults = .standard) {
        let sanitized = info.sanitized()
        userDefaults.set(sanitized.bloodType, forKey: bloodKey)
        userDefaults.set(sanitized.allergies, forKey: allergiesKey)
        userDefaults.set(sanitized.medication, forKey: medicationKey)
        userDefaults.set(sanitized.notes, forKey: notesKey)
    }
}

private extension String {
    var isBlank: Bool {
        trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }
}

File: Android/app/src/main/java/com/auralis/crisisconnect/data/local/MedicalInfoStore.kt
package com.auralis.crisisconnect.data.local

import android.content.Context
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import com.auralis.crisisconnect.settingsDataStore
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map

/**
 * The user's optional emergency medical details (blood type, allergies, medication, note).
 *
 * Deliberately LOCAL + BLE-ONLY: the fields ride the encrypted SOS peer_info to a verified
 * rescuer during an emergency and are never uploaded to the cloud. Empty fields are never sent.
 */
data class MedicalInfo(
    val bloodType: String = "",
    val allergies: String = "",
    val medication: String = "",
    val notes: String = ""
) {
    val isEmpty: Boolean
        get() = bloodType.isBlank() && allergies.isBlank() &&
            medication.isBlank() && notes.isBlank()

    fun sanitized(): MedicalInfo = MedicalInfo(
        bloodType = bloodType.trim().take(MedicalInfoStore.MAX_BLOOD_LENGTH),
        allergies = allergies.trim().take(MedicalInfoStore.MAX_FIELD_LENGTH),
        medication = medication.trim().take(MedicalInfoStore.MAX_FIELD_LENGTH),
        notes = notes.trim().take(MedicalInfoStore.MAX_FIELD_LENGTH)
    )
}

object MedicalInfoStore {
    const val MAX_BLOOD_LENGTH = 8
    const val MAX_FIELD_LENGTH = 200

    private val BLOOD_KEY = stringPreferencesKey("medical_blood_type")
    private val ALLERGIES_KEY = stringPreferencesKey("medical_allergies")
    private val MEDICATION_KEY = stringPreferencesKey("medical_medication")
    private val NOTES_KEY = stringPreferencesKey("medical_notes")

    fun observe(context: Context): Flow<MedicalInfo> {
        return context.applicationContext.settingsDataStore.data.map { prefs ->
            MedicalInfo(
                bloodType = prefs[BLOOD_KEY] ?: "",
                allergies = prefs[ALLERGIES_KEY] ?: "",
                medication = prefs[MEDICATION_KEY] ?: "",
                notes = prefs[NOTES_KEY] ?: ""
            )
        }
    }

    suspend fun save(context: Context, info: MedicalInfo) {
        val sanitized = info.sanitized()
        context.applicationContext.settingsDataStore.edit { prefs ->
            prefs[BLOOD_KEY] = sanitized.bloodType
            prefs[ALLERGIES_KEY] = sanitized.allergies
            prefs[MEDICATION_KEY] = sanitized.medication
            prefs[NOTES_KEY] = sanitized.notes
        }
    }
}
